File: converter/pdf/pdf.cpp
#include "pdf.h"
#include <fstream>
#include <filesystem>
#include <iostream>

using namespace std;
namespace fs = filesystem;

Pdf::Pdf(string name) {
    this->name = name;
    this->path = "generated_pdfs/" + name;

    fs::create_directory(this->path);

    // Read the amount of files within the path
    for (const auto & entry : fs::directory_iterator(path)) {
        // Add file path to private class vector
        string file = entry.path().string();
        this->currentFiles.push_back(file);
    }
}

Pdf::~Pdf() {}

void Pdf::build() {
    // Check if we have any files to combine.
    if (this->currentFiles.size() >= 1) {
        combineFiles();
    }
    // The command to convert html to pdf
    string command = "wkhtmltopdf " + this->path + "/pdf.html " + "generated_pdfs/" + this->name + ".pdf";
    // Now execute the command line
    system(command.c_str());
    // Now Let's remove the working pdf directory
    fs::remove(this->path + "/pdf.html");
    fs::remove(this->path);
}

void Pdf::addFile(string path) {
    // Lets add the file to the genereated_pdfs path
    // First open file
    ifstream htmlFile(path);
    // Data holder
    string data;
    vector<string> htmlData;

    while (getline(htmlFile, data)) {
        htmlData.push_back(data);
    }
    htmlFile.close();

    string fileForPdfPath = this->path + "/" + to_string(this->currentFiles.size()) + ".html";

    // Now to write the new file.
    ofstream fileForPdf(fileForPdfPath);
    // Write to file
    for (auto str : htmlData) {
        fileForPdf << str << endl;
    }
    // Close file
    fileForPdf.close();

    // Add file path to files
    this->currentFiles.push_back(fileForPdfPath);
}

void Pdf::combineFiles() {
    // The file to write to
    ofstream pdf(this->path + "/pdf.html");
    // Lets loop through the current files
    for (auto filePath : this->currentFiles) {
        // Open the file
        ifstream file(filePath);
        string data;
        // Now write to pdf.html
        while(getline(file, data)) {
            pdf << data;
        }
        file.close();

        try {
            // Remove
            fs::remove(filePath);
        } catch(const fs::filesystem_error) {
            cout << "Could not delete file " << filePath << " from path!" << endl;
            continue;
        }
    }
    pdf.close();
}

File: backend/converter/pdf.py
from backend.data.dir import manga_dir, novels_dir
import subprocess
import os
from pathlib import Path
from glob import glob
import codecs


NOVEL_PDF = 0
MANGA_PDF = 1


class PDF:
    def __init__(self, pdf_name, pdf_type):
        self.name = pdf_name
        
        if pdf_type == NOVEL_PDF:
            self.path = novels_dir() + self.name + "-pdf"
        elif pdf_type == MANGA_PDF:
            self.path = manga_dir() + self.name + "-pdf"

        self.files = glob(f"{self.path}/*.html")

        # Check if the path to the PDF exists or not
        if not Path(self.path).exists():
            # Create it
            os.mkdir(self.path)

    def add_html(self, html_path):
        """
        Add a html file to the PDF. Used for novels.

        Params:
            - <html_path: str> The path to the HTML file
        """
        # Add html_data to a new file within PDF directory
        html_data = ""
        with codecs.open(html_path, "r", "utf-8") as html_file:
            html_data = html_file.read()

        # Write to new file
        new_file_path = self.path + f"/{len(self.files)}_.html"
        with codecs.open(new_file_path, 'w', 'utf-8') as new_file:
            for data in html_data:
                new_file.write(data)

        # Add new file path to self.files
        self.files.append(new_file_path)

    def add_image(self, image_path):
        """
        Add a Image to the PDF. Used for Manga.

        Params:
            - <image_path: str> The path to the Image file.
        """
    
    def build(self):
        """
        Build the PDF.

        Returns: <pdf_path: str>
        """
        # Check if we have more than one file
        if len(self.files) > 1:
            self.__combine_files()
        
        finished_pdf_path = f"{self.path}/{self.name}.pdf"

        command = f"wkhtmltopdf {self.path}/pdf.html {finished_pdf_path}"
        subprocess.call(command, shell=True)

        # Now remove the pdf.html
        os.unlink(f"{self.path}/pdf.html")

        return finished_pdf_path

    def __combine_files(self):
        """
        Combine the files into one file. 
        """
        # Open the file to write to
        with codecs.open(self.path + "/pdf.html", 'w', 'utf-8') as pdf_html:
            # Let's loop through the current files
            for f in self.files:
                with codecs.open(f, 'r', 'utf-8') as html:
                    # Write to the pdf_html
                    pdf_html.write(html.read())
                # Now delete the file
                os.unlink(f)
            # Daijobu!